html_xblock/static/js/html.js

```diff
@@ -0,0 +1,267 @@
+      });
+      $wrapper.removeClass('is-set');
+      $resetButton.removeClass('active').addClass('inactive');
+    });
+  });
+
+  return fields;
+}
+
+function HTML5XBlock(runtime, element, data) {
+  document.getElementById("default-tab").click();  // Will open the XBlock by showing the default tab
+
+  const editor = configureTheEditor(data);
+  var fields = extractXBlockFields();
+  var elements;
+
+  function studioSubmit() {
+    const ContentHandlerUrl = runtime.handlerUrl(element, "update_content");
+    const SettingsHandlerUrl = runtime.handlerUrl(element, "submit_studio_edits");
+    const content = (data.editor === "visual") ? tinymce.get("html5-textarea").getContent() : editor.getValue();
+    const fields_data = getSettingsValues(fields);
+    var errorMessage = "This may be happening because of an error with our server or your internet connection. Try refreshing the page or making sure you are online.";
+
+    runtime.notify('save', {state: 'start', message: "Saving"});
+    $.ajax({
+      type: "POST",
+      url: SettingsHandlerUrl,
+      data: JSON.stringify(fields_data),
+      dataType: "json",
+      global: false,  // Disable Studio's error handling that conflicts with studio's notify('save') and notify('cancel') :-/
+      success: function (response) {
+        $.ajax({
+          type: "POST",
+          url: ContentHandlerUrl,
+          data: JSON.stringify({"content": content}),
+          dataType: "json",
+          global: false,  // Disable Studio's error handling that conflicts with studio's notify('save') and notify('cancel') :-/
+          success: function (response) {
+            runtime.notify('save', {state: 'end'});
+          }
+        }).fail(function (jqXHR) {
+          runtime.notify('error', {title: "Unable to update content", message: errorMessage});
+        })
+      }
+    }).fail(function (jqXHR) {
+      if (jqXHR.responseText) { // Is there a more specific error message we can show?
+        try {
+          errorMessage = JSON.parse(jqXHR.responseText).error;
+          if (typeof errorMessage === "object" && errorMessage.messages) {
+            // e.g. {"error": {"messages": [{"text": "Unknown user 'bob'!", "type": "error"}, ...]}} etc.
+            errorMessage = $.map(errorMessage.messages, function (msg) {
+              return msg.text;
+            }).join(", ");
+          }
+        } catch (error) {
+          errorMessage = jqXHR.responseText.substr(0, 300);
+        }
+      }
+      runtime.notify('error', {title: "Unable to update settings", message: errorMessage});
+    })
+  }
+
+  elements = document.querySelectorAll(".save-button");
+  Array.prototype.forEach.call(elements, function (elem) {
+    elem.addEventListener("click", function (event) {
+      event.preventDefault();
+      studioSubmit();
+    });
+  });
+
+  elements = document.querySelectorAll(".cancel-button");
+  Array.prototype.forEach.call(elements, function (elem) {
```

Comment:
If you want, this can be simplified.

```javascript
element = typeof element[0] === 'undefined' ? element : element[0]; // Fix: sometimes edX passes a jQuery element, other times a DOM element

const addClickFn = function(el, fn) { el.addEventListener("click", function (event) { event.preventDefault(); fn(event); } };

addClickFn(element.querySelector('.save-button'), studioSubmit);
addClickFn(element.querySelector('.cancel-button'), function() { runtime.notify('cancel', {}); });
```

It's better to use element.querySelector() because it's scope to this XBlock, compared to document.querySelector which could find other buttons belonging to other XBlocks.

Reply:
Thanks. I got to say, debugging and working with JQuery and javascript took a lot of time here, I've never wrote a completed module from scratch. Your comments are always appreciated as they help me continue learning.